src/Backups/BackupInfo.cpp

```diff
@@ -98,4 +99,23 @@ String BackupInfo::toStringForLogging() const
     return toAST()->formatForLogging();
 }
 
+void BackupInfo::copyS3CredentialsTo(BackupInfo & dest) const
+{
+    /// named_collection case, no need to update
+    if (!dest.id_arg.empty() || !id_arg.empty())
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "use_same_s3_credentials_for_base_backup is not compatible with named_collections");
+
+    auto & dest_args = dest.args;
+
+    if ((dest_args.size() != 1) && (dest_args.size() != 3))
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Backup S3 requires 1 or 3 arguments: url, [access_key_id, secret_access_key]");
```

Comment:
It's not a logical error, a logical error means "that cannot happen". In debug version the server even stops on a logical error.

Reply:
Yes, and here it should never happen, since the same validation exists during parsing backup query, hence LOGICAL_ERROR.

Reply:
https://github.com/ClickHouse/ClickHouse/blob/55af6e5c3e1bfdee95845279f33a265a9343dca7/src/Backups/registerBackupEngineS3.cpp#L79-L81

Reply:
>It seems copyS3CredentialsTo is called before creator_fn in registerBackupEngineS3.cpp

Not sure that I got this comment. Can you please elaborate?

Reply:
I meant `BackupImpl::open()` calls `copyS3CredentialsTo()` before calling `BackupFactory::instance().createBackup()` for the base backup.

Reply:
Yes, and it should be like that, otherwise it will use incorrect credentials. Not sure that I follow your point.

Reply:
I guess I got your point about calling `copyS3CredentialsTo` after `createBackup`, please correct me if I'm wrong.

It may throw `LOGICAL_ERROR` in this case for number of arguments check. The latest version of the patch checks number of arguments only for the backup itself (not the `base_backup`), so now, it should not be a problem.

Reply:
Anyway I think logical error is a too big error for that arguments checking. And it seems better not to put any deep knowledge into this function `copyS3CredentialsTo()` about what and where it's called. `BAD_ARGUMENTS` must be enough for all the checking in this function.

Reply:
OK